### PR TITLE
Extend Thrift codegen for auth.actor_uuid annotation to support nested structs/typedefs, cycles and cross-package imports

### PR DESCRIPTION
**Pull request by mark200**
- [ ] Description and context for reviewers:
  - Plugin **thriftrw-plugin-yarpc** extends _auth.actor_uuid_ annotation to support nested structs + typedefs + cross imports. This change completes the basic requirements, which are:
    -  `auth.actor_uuid` needs to exist only on a field
    -  `auth.actor_uuid` can only be applied to a string/typedef-of-string field.
    - From every method arguments list, we should have a path from methods_args up to the annotated field
    
  For example, the following thrift file can now extract the annotation and emit an ActorUUID accessor for each method in `TestService`. **For cycles this would only work when the annotation does not exist on the cycle branch.**

    ```thrift

    struct Inner {
        1: required string id (auth.actor_uuid = "true")
    }
    
    typedef Inner StructAlias

    struct CycleNode {
      1: required CycleNode loop
      2: optional string id (auth.actor_uuid = "true")
    }
    
    service TestService {
      string testMethodTypedef(
          1: string firstUUID,
          2: StructAlias second,
      )
      string testMethodCycle(
          1: CycleNode arg,
       )
    }
    ```
    - Another addition is that the method signature of ActorUUID is now changed to return a slice of strings.

RELEASE NOTES:
Extend Thrift codegen for annotation `_auth.actor_uuid_` to support nested typedefs/structs, cycles and cross-package imports.

### CODE REVIEW

File: encoding/thrift/thriftrw-plugin-yarpc/uuid.go [hunk 1]
```diff
@@ -69,108 +82,172 @@ func yarpcUUIDGenerator(data *moduleTemplateData, files map[string][]byte) error
...
-		return hits[0]
+// castImportPath returns the Go import path that owns the struct
+// the chain has to cast through, or "" when that package is the
+// same as the file being generated (no qualifier needed).
+func castImportPath(ss *compile.StructSpec, data *moduleTemplateData) string {
+	if ss == nil || data == nil || data.Module == nil {
+		return ""
 	}
-	return nil
+	importPath := data.importPathForThriftFile(ss.ThriftFile())
```
Comment (bananacocodrilo) [inline encoding/thrift/thriftrw-plugin-yarpc/uuid.go:221]: Does ss.ThriftFile match  req.Modules[].ThriftFilePath?

Can you add cross-package imports to WITHSERVICES.thrift? A file in a different path that contains an annotated typedef.

File: encoding/thrift/thriftrw-plugin-yarpc/uuid.go [hunk 2]
```diff
@@ -69,108 +82,172 @@ func yarpcUUIDGenerator(data *moduleTemplateData, files map[string][]byte) error
 	// Plugins do not have access to it via the API, so it is compiled once in
 	// Generate and shared across generators; this generator writes extra
 	// methods to types annotated with auth.actor_uuid in that Thrift file.
-	compiledModule := data.CompiledModule
-
-	// Pre-flight pass over the same structures the template iterates,
-	// enforcing the at-most-one-annotation-per-container invariant.
-	// The file itself is always emitted (even when
-	// no annotated fields exist, the output is a header-only package file)
-	if err := validateUUIDAnnotations(compiledModule); err != nil {
```
Comment (bananacocodrilo) [inline encoding/thrift/thriftrw-plugin-yarpc/uuid.go:78]: Why don't we validate this anymore? Are we dropping some of the fields silently?

File: encoding/thrift/thriftrw-plugin-yarpc/uuid.go [hunk 3]
```diff
@@ -195,55 +273,17 @@ func serviceHasActorUUIDMethod(svc *api.Service, mod *compile.Module) bool {
...
-			}
-		}
-	}
-
-	return nil
-}
-
-// tooManyAnnotations returns an error naming the first two hits when
-// more than one annotation was found in a single container.
-func tooManyAnnotations(hits []*uuidArg, ownerLabel string) error {
```
Comment (bananacocodrilo) [inline encoding/thrift/thriftrw-plugin-yarpc/uuid.go:234]: why are we not doing this anymore?

File: encoding/thrift/thriftrw-plugin-yarpc/internal/tests/WITHSERVICES/types_yarpc_uuid.go [hunk 4]
```diff
@@ -5,30 +5,42 @@ package WITHSERVICES
 
 // ActorUUID returns the UUID string from the field annotated with
 // auth.actor_uuid, or "" if it is unset.
-func (t *Struct) ActorUUID() string {
```
Comment (bananacocodrilo) [inline encoding/thrift/thriftrw-plugin-yarpc/internal/tests/WITHSERVICES/types_yarpc_uuid.go:8]: Why are we removing this public method?

File: encoding/thrift/thriftrw-plugin-yarpc/internal/tests/WITHSERVICES.thrift [hunk 5]
```diff
@@ -18,6 +19,53 @@ struct StructRequiredUUID {
...
+struct Inner {
+    1: optional string innerUUID (auth.actor_uuid = "true")
+}
+
+typedef Inner AliasedInner
+typedef AliasedInner DoubleAliasedInner
+
+struct OuterWithAlias {
+    1: optional AliasedInner inner
+    2: optional DoubleAliasedInner deeplyAliased
```
Comment (bananacocodrilo) [inline encoding/thrift/thriftrw-plugin-yarpc/internal/tests/WITHSERVICES.thrift]: This should be failing with two fields with the annotation no?

File: encoding/thrift/thriftrw-plugin-yarpc/internal/tests/WITHSERVICES.thrift [hunk 6]
```diff
@@ -18,6 +19,53 @@ struct StructRequiredUUID {
...
+// must cast through Inner to reach GetXxx() accessors. The two-hop
+// alias confirms the cast stays valid even after multiple typedefs
+// because `*DoubleAliasedInner`, `*AliasedInner` and `*Inner` all
+// share the same underlying struct type.
+struct Inner {
+    1: optional string innerUUID (auth.actor_uuid = "true")
+}
+
+typedef Inner AliasedInner
+typedef AliasedInner DoubleAliasedInner
```
Comment (bananacocodrilo) [inline encoding/thrift/thriftrw-plugin-yarpc/internal/tests/WITHSERVICES.thrift:62]: Can we have the opposite scenario too?


```
struct Inner {
    1: optional string innerUUID 
}
typedef Inner AliasedInner
typedef AliasedInner DoubleAliasedInner (auth.actor_uuid = "true")
```

File: encoding/thrift/thriftrw-plugin-yarpc/uuid.go (hunk 3 above)
Comment (mark200) [reply encoding/thrift/thriftrw-plugin-yarpc/uuid.go:234]: Part of previous implementation that throws error when we find a second annotated field in the tree

File: encoding/thrift/thriftrw-plugin-yarpc/internal/tests/WITHSERVICES/types_yarpc_uuid.go (hunk 4 above)
Comment (mark200) [reply encoding/thrift/thriftrw-plugin-yarpc/internal/tests/WITHSERVICES/types_yarpc_uuid.go:8]: This is an accessor for the type itself, and will not be needed. The server codegen only relies on `service_ method_args`'s `ActorUUID` accessor.

File: encoding/thrift/thriftrw-plugin-yarpc/internal/tests/WITHSERVICES.thrift (hunk 5 above)
Comment (mark200) [reply encoding/thrift/thriftrw-plugin-yarpc/internal/tests/WITHSERVICES.thrift]: This is an edge case, and it just proves that the emitted `ActorUUID` accessor would be pointing to the first found annotated field, i.e. `AliasedInner -> Inner -> innerUUID`

File: encoding/thrift/thriftrw-plugin-yarpc/internal/tests/WITHSERVICES.thrift (hunk 6 above)
Comment (mark200) [reply encoding/thrift/thriftrw-plugin-yarpc/internal/tests/WITHSERVICES.thrift:62]: The requirement is only for a field (of a `Struct` of `args`) to be annotated

File: encoding/thrift/thriftrw-plugin-yarpc/uuid.go (hunk 1 above)
Comment (mark200) [reply encoding/thrift/thriftrw-plugin-yarpc/uuid.go:221]: Done, it's method testImportedTypedef

File: encoding/thrift/thriftrw-plugin-yarpc/uuid.go (hunk 3 above)
Comment (mark200) [reply encoding/thrift/thriftrw-plugin-yarpc/uuid.go:234]: Update - validation logic has been rewritten

File: encoding/thrift/thriftrw-plugin-yarpc/internal/tests/WITHSERVICES/types_yarpc_uuid.go (hunk 4 above)
Comment (bananacocodrilo) [reply encoding/thrift/thriftrw-plugin-yarpc/internal/tests/WITHSERVICES/types_yarpc_uuid.go:8]: 👍 
Let's make sure we mention this and the ActorUUIDValidator changes in the release notes

File: encoding/thrift/thriftrw-plugin-yarpc/uuid.go [hunk 7]
```diff
@@ -90,87 +103,170 @@ func yarpcUUIDGenerator(data *moduleTemplateData, files map[string][]byte) error
...
+// it must carry the annotation and its type must be a plain string or
+// a single-hop typedef of string.
+func matchUUIDLeaf(f *compile.FieldSpec) *uuidPath {
+	if f == nil {
+		return nil
+	}
+	if _, ok := f.Annotations[_UUIDAnnotationKey]; !ok {
+		return nil
+	}
+	switch t := f.Type.(type) {
```
Comment (bananacocodrilo) [inline encoding/thrift/thriftrw-plugin-yarpc/uuid.go:238]:
```
t := f.Type
    isTypedef := false
    for {
        switch tt := t.(type) {
        case *compile.StringSpec:
            return &uuidPath{Steps: []uuidPathStep{{Field: f}}, IsTypedef: isTypedef}
        case *compile.TypedefSpec:
            isTypedef = true
            t = tt.Target
        default:
            return nil
        }
    }
```

File: encoding/thrift/thriftrw-plugin-yarpc/uuid_test.go [hunk 8]
```diff
@@ -31,149 +32,220 @@ import (
...
 		require.NoError(t, err)
-		assert.NoError(t, validateUUIDAnnotations(spec))
+		fn, ok := spec.Services["TestService"].Functions["testMethod"]
+		require.True(t, ok)
+
+		got := uuidPathInArgs(fn, nil)
+		require.Len(t, got, 1)
+		assert.Equal(t, []string{"identifier"}, stepNames(got[0]))
+		assert.True(t, got[0].IsTypedef)
 	})
```
Comment (bananacocodrilo) [inline encoding/thrift/thriftrw-plugin-yarpc/uuid_test.go:100]: t.Run("doubleTypedefStringLeafAnnotation", func(t *testing.T) {
		spec, err := compile.Compile("internal/uuid_test/multiHopTypedefStringLeaf.thrift")
		require.NoError(t, err)
		fn, ok := spec.Services["TestService"].Functions["testMethod"]
		require.True(t, ok)

		got := uuidPathInArgs(fn, nil)
		require.Len(t, got, 1, "expected one path for the doubly-aliased string leaf")
		assert.True(t, got[0].IsTypedef)
	})

Comment (bananacocodrilo) [reply encoding/thrift/thriftrw-plugin-yarpc/uuid_test.go:100]: DoubleAlias is typedef UUIDBase, which is typedef string. matchUUIDLeaf only peels one typedef, so it sees TypedefSpec→TypedefSpec instead of TypedefSpec→StringSpec and returns nil

**Review by bananacocodrilo — APPROVED** [review]
